src/dataset.jl

```diff
@@ -1,34 +1,33 @@
 """
-    KeyedDataset{T}
+    KeyedDataset{K, T}
 
 A `KeyedDataset` describes an associative collection of component `KeyedArray`s with constraints
 on their shared dimensions.
 
 # Fields
 - `constraints::OrderedSet{Pattern}` - Constraint [`Pattern`](@ref)s on shared dimensions.
-- `data::LittleDict{Tuple{Vararg{Symbol}}, T}` - Flattened key paths as tuples of symbols
-  to each component array of type `T`.
+- `data::LittleDict{K, T}` - Flattened key paths as type `K <: Tuple` to each component array of type `T`.
 """
-@auto_hash_equals struct KeyedDataset{T<:XArray}
+@auto_hash_equals struct KeyedDataset{K<:Tuple, T<:XArray}
```

Comment:
We're parameterizing by the tuple key type to minimize performance issues (ie: `Tuple` isn't a concrete type).